controller/src/main/java/org/jboss/as/controller/operations/global/ReadFeatureHandler.java

```diff
@@ -0,0 +1,1179 @@
+import org.jboss.as.controller.registry.AttributeAccess;
+
+/**
+ * Global operation handler that describes the resource as a Galleon feature.
+ *
+ * @author <a href="kabir.khan@jboss.com">Kabir Khan</a>
+ * @author Brian Stansberry (c) 2012 Red Hat Inc.
+ * @author Emmanuel Hugonnet
+ * @author Alexey Loubyansky
+ */
+public class ReadFeatureHandler extends GlobalOperationHandlers.AbstractMultiTargetHandler {
+
+    public static final OperationDefinition DEFINITION = new SimpleOperationDefinitionBuilder(READ_FEATURE_OPERATION, ControllerResolver.getResolver("global"))
+            .setParameters(RECURSIVE, RECURSIVE_DEPTH)
+            .setReadOnly()
+            .setReplyValueType(ModelType.OBJECT)
+            .build();
+
+    final ImmutableCapabilityRegistry capabilityRegistry;
+
+    private static final String $PROFILE_PREFIX = "$profile.";
+    private static final String ADDRESS_PARAMETERS = "addr-params";
+    private static final String ADDRESS_PARAMETERS_MAPPING = "addr-params-mapping";
+    private static final String DOMAIN_EXTENSION = "domain.extension";
+    private static final String FEATURE_ID = "feature-id";
+    private static final String HOST_EXTENSION = "host.extension";
+    private static final String OPERATION_PARAMETERS = "op-params";
+    private static final String OPERATION_PARAMETERS_MAPPING = "op-params-mapping";
+    private static final String PARAMETERS = "params";
+    private static final String REFERENCES = "refs";
+
+    //Placeholder for NoSuchResourceExceptions coming from proxies to remove the child in ReadFeatureHandler
+    private static final ModelNode PROXY_NO_SUCH_RESOURCE;
+
+    static {
+        //Create something non-used since we cannot
+        ModelNode none = new ModelNode();
+        none.get("no-such-resource").set("no$such$resource");
+        none.protect();
+        PROXY_NO_SUCH_RESOURCE = none;
+    }
+
+    public static OperationStepHandler getInstance(ImmutableCapabilityRegistry capabilityRegistry) {
+        return new ReadFeatureHandler(capabilityRegistry);
+    }
+
+    private ReadFeatureHandler(final ImmutableCapabilityRegistry capabilityRegistry) {
+        super(true);
+        this.capabilityRegistry = capabilityRegistry;
+    }
+
+    ReadFeatureAccessControlContext getAccessControlContext() {
+        return null;
+    }
+
+    @Override
+    void doExecute(OperationContext context, ModelNode operation, FilteredData filteredData, boolean ignoreMissingResource) throws OperationFailedException {
+        final PathAddress address = context.getCurrentAddress();
+        ReadFeatureAccessControlContext accessControlContext = getAccessControlContext() == null ? new ReadFeatureAccessControlContext(address, null) : getAccessControlContext();
+        doExecute(context, operation, accessControlContext);
+    }
+
+    void doExecute(OperationContext context, ModelNode operation, ReadFeatureAccessControlContext accessControlContext) throws OperationFailedException {
```

Comment:
This should be private.

Reply:
Apparently it can't be private. Otherwise it won't build.